### PR TITLE
Re-enable nav2_costmap_2d warnings

### PR DESCRIPTION
**Pull request by bpwilcox**
## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #535 |

## Description of contribution in a few bullet points
- Re-enables and passes warnings in nav2_costmap_2d

### CODE REVIEW

File: nav2_costmap_2d/src/costmap_2d_cloud.cpp [hunk 1]
```diff
@@ -154,7 +154,7 @@ void voxelCallback(const nav2_msgs::msg::VoxelGrid::ConstSharedPtr grid)
       // uint32_t a = g_colors_a[c.status] * 255.0;
 
       uint32_t col = (r << 16) | (g << 8) | b;
-      cval = *reinterpret_cast<float *>(&col);
+      memcpy(&cval, &col, sizeof col);
```
Comment (crdelsey) [inline nav2_costmap_2d/src/costmap_2d_cloud.cpp]: I don't think this is valid. Most likely cval never had memory allocated to copy into. You are also changing behavior here; adding work that probably isn't necessary.

What was the compiler warning?

File: nav2_costmap_2d/src/costmap_2d_cloud.cpp [hunk 2]
```diff
@@ -185,7 +185,7 @@ void voxelCallback(const nav2_msgs::msg::VoxelGrid::ConstSharedPtr grid)
...
```
Comment (crdelsey) [inline nav2_costmap_2d/src/costmap_2d_cloud.cpp]: Same here. I doubt `cval` is valid, and it is definitely undesirable to do the extra memcpy.

File: nav2_costmap_2d/test/integration/inflation_tests.cpp [hunk 3]
```diff
@@ -224,15 +224,15 @@ TEST_F(TestNode, testInflationShouldNotCreateUnknowns)
...
+  addInflationLayer(layers, tf, node_);
   layers.setFootprint(polygon);
 
   addObservation(olayer, 0, 0, MAX_Z);
 
   layers.updateMap(0, 0, 0);
   nav2_costmap_2d::Costmap2D * costmap = layers.getCostmap();
 
-  EXPECT_EQ(countValues(*costmap, nav2_costmap_2d::NO_INFORMATION), 0);
+  EXPECT_EQ(countValues(*costmap, nav2_costmap_2d::NO_INFORMATION), (unsigned int)0);
```
Comment (crdelsey) [inline nav2_costmap_2d/test/integration/inflation_tests.cpp]:
```suggestion
  EXPECT_EQ(countValues(*costmap, nav2_costmap_2d::NO_INFORMATION), 0u);
```

File: nav2_costmap_2d/test/unit/array_parser_test.cpp [hunk 4]
```diff
@@ -38,9 +38,9 @@ TEST(array_parser, basic_operation)
   std::string error;
   std::vector<std::vector<float>> vvf;
   vvf = nav2_costmap_2d::parseVVF("[[1, 2.2], [.3, -4e4]]", error);
-  EXPECT_EQ(2, vvf.size() );
-  EXPECT_EQ(2, vvf[0].size() );
-  EXPECT_EQ(2, vvf[1].size() );
+  EXPECT_EQ((unsigned int)2, vvf.size() );
```
Comment (crdelsey) [inline nav2_costmap_2d/test/unit/array_parser_test.cpp]:
```suggestion
  EXPECT_EQ(2u, vvf.size() );
```

Same for the other ones below

File: nav2_costmap_2d/src/costmap_2d_cloud.cpp (hunk 1 above)
Comment (crdelsey) [reply nav2_costmap_2d/src/costmap_2d_cloud.cpp]: Never mind. I went through the code to understand the context, and I think your approach is good.

File: nav2_costmap_2d/src/costmap_2d_cloud.cpp (hunk 2 above)
Comment (crdelsey) [reply nav2_costmap_2d/src/costmap_2d_cloud.cpp]: Ditto. I take my comment back. :-)

File: nav2_costmap_2d/test/integration/inflation_tests.cpp (hunk 3 above)
Comment (bpwilcox) [reply nav2_costmap_2d/test/integration/inflation_tests.cpp]: Forgot about that syntax :)

File: nav2_costmap_2d/src/costmap_2d_cloud.cpp (hunk 1 above)
Comment (bpwilcox) [reply nav2_costmap_2d/src/costmap_2d_cloud.cpp]: It was warning about strict aliasing. I found this method of handling it.

File: nav2_costmap_2d/plugins/static_layer.cpp [hunk 5]
```diff
@@ -273,7 +273,7 @@ StaticLayer::incomingUpdate(map_msgs::msg::OccupancyGridUpdate::ConstSharedPtr u
 
 void
 StaticLayer::updateBounds(
-  double robot_x, double robot_y, double robot_yaw, double * min_x,
+  double /*robot_x*/, double /*robot_y*/, double /*robot_yaw*/, double * min_x,
```
Comment (mjeronimo) [inline nav2_costmap_2d/plugins/static_layer.cpp]: If these arguments are not used, how about removing them from the argument list? Can be outside the scope of the current PR, but something to consider.

File: nav2_costmap_2d/include/nav2_costmap_2d/obstacle_layer.hpp [hunk 6]
```diff
@@ -44,7 +44,10 @@
 
 #include "rclcpp/rclcpp.hpp"
 #include "laser_geometry/laser_geometry.hpp"
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wreorder"
```
Comment (mjeronimo) [inline nav2_costmap_2d/include/nav2_costmap_2d/obstacle_layer.hpp]: Let's submit a PR to geometry2/tf2_ros to address this issue. A simple fix for message_filter.h and we can remove the pragmas.

Comment (bpwilcox) [reply nav2_costmap_2d/include/nav2_costmap_2d/obstacle_layer.hpp]: Agreed. I'll do that next week.

File: nav2_costmap_2d/plugins/static_layer.cpp (hunk 5 above)
Comment (bpwilcox) [reply nav2_costmap_2d/plugins/static_layer.cpp]: this function is from the virtual base class, so I'm not sure that is allowed

**Pull request merged**